src/modules/tpm/src/lib/Tpm.cpp

```diff
@@ -49,37 +51,82 @@ int Tpm::Get(const char* objectName, MMI_JSON_STRING* payload, int* payloadSizeB
     int status = MMI_OK;
     std::string data;
 
-    if (std::strcmp(objectName, TPM_STATUS) == 0)
-    {
-        GetStatus(data);
-    }
-    else if (std::strcmp(objectName, TPM_VERSION) == 0)
-    {
-        GetVersion(data);
-    }
-    else if (std::strcmp(objectName, TPM_MANUFACTURER) == 0)
+    if ((std::strcmp(objectName, TPM_STATUS) == 0) || (std::strcmp(objectName, TPM_VERSION) == 0) || (std::strcmp(objectName, TPM_MANUFACTURER) == 0))
     {
-        GetManufacturer(data);
+        if (std::strcmp(objectName, TPM_STATUS) == 0)
+        {
+            GetStatus(data);
+        }
+        else if ((std::strcmp(objectName, TPM_VERSION) == 0) && (this->m_hasCapabilitiesFile))
+        {
+            GetVersionFromCapabilitiesFile(data);
+        }
+        else if ((std::strcmp(objectName, TPM_MANUFACTURER) == 0) && (this->m_hasCapabilitiesFile))
+        {
+            GetManufacturerFromCapabilitiesFile(data);
+        }
     }
     else
     {
+        if (IsFullLoggingEnabled())
+        {
+            OsConfigLogError(TpmLog::Get(), "Invalid argument, objectName %s not found", objectName);
+        }
         status = EINVAL;
     }
 
+    if ((2 >= data.length()) && (status == MMI_OK) && ((0 == std::strcmp(objectName, TPM_VERSION)) || (0 == std::strcmp(objectName, TPM_MANUFACTURER))))
+    {
+        this->m_hasCapabilitiesFile = false;
+        std::string tpmProperty;
+        status = Tpm2Utils::GetTpmPropertyFromDeviceFile(objectName, tpmProperty);
+        if ((status == MMI_OK) && (!tpmProperty.empty()))
+        {
+            if (0 == std::strcmp(objectName, TPM_VERSION))
+            {
+                std::regex re(g_tpmVersionFromDeviceFile);
+                std::smatch match;
+                if (std::regex_search(tpmProperty, match, re))
+                {
+                    data = char('"') + tpmProperty + char('"');
```

Comment:
>'"' [](http://example.com/codeflow?start=32&length=3)

Define a constant for this charecter and use it

Reply:
Addressed in the latest iteration.